src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs

```diff
@@ -811,12 +811,12 @@ private ISet<string> CalculateFilesToExcludeInPack(IPackTaskRequest<IMSBuildItem
                         continue;
                     }
 
-                    var versionToUse = targetLibrary.Version;
+                    var versionToUse = new VersionRange(targetLibrary.Version);
```

Comment:
This seems like an idempotent change to me. 
It just creates the VersionRange model earlier. 

Am I missing something?

Reply:
This is to change type of `versionToUse` to `VersionRange` only.
The critical part is the replacement of `NuGetVersion.Parse` with `VersionRange.Parse`.

That is that `request.ProjectReferencesWithVersions` can contain version ranges.
Then I can force exact versions for project dependencies in my project files:
```xml
  <Target Name="_ExactProjectReferencesVersion" AfterTargets="_GetProjectReferenceVersions">
    <ItemGroup>
      <_ProjectReferencesWithExactVersions Include="@(_ProjectReferencesWithVersions)">
        <ProjectVersion>[%(_ProjectReferencesWithVersions.ProjectVersion)]</ProjectVersion>
      </_ProjectReferencesWithExactVersions>
    </ItemGroup>

    <ItemGroup>
      <_ProjectReferencesWithVersions Remove="@(_ProjectReferencesWithVersions)" />
      <_ProjectReferencesWithVersions Include="@(_ProjectReferencesWithExactVersions)" />
    </ItemGroup>
  </Target>
```

Reply:
Makes sense. 
That intention wasn't obvious to me :)

Reply:
I am comfortable with this change, as an extensibility point. At least until the mega issue gets a resolution. 

Can you please add a test for this? 
https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs is the test collection. 

Note that you need to run build.ps1 to pack  the nupkgs first in order to run those tests.

Reply:
The test `fPackCommand_PackProject_GetsProjectRefVersionFromMsbuild` verifies that I haven't broken anything.

Shall I add a test with the MSBuild code snipped from my last comment?
I think the test will break when the MSBuild scripts are updated as it relies on internals (all properties and targets start with an underscore).

Reply:
> I think the test will break when the MSBuild scripts are updated as it relies on internals 

And that's ok. 

I think we should have that test in order to ensure that other changes that are not https://github.com/NuGet/Home/issues/5556 don't break it accidentally.

Reply:
I've added the test.

But I was not able to compile without errors on my machine. So I'm not sure if everything is working as intended.